20250404-RS-workflow-rerun-from-activity.md

```diff
@@ -0,0 +1,142 @@
+
+If the targeted `eventID` does not exist, or is not an Activity event, the API will return an error to the client.
+
+```proto
+service TaskHubSidecarService {
+    // Rerun a Workflow from a specific event ID from an activity.
+    rpc RerunWorkflowFromActivity(RerunWorkflowFromActivityRequest) returns (RerunWorkflowFromActivityResponse);
+}
+
+// RerunWorkflowFromActivityRequest is used to rerun a workflow instance from a
+// specific event ID.
+message RerunWorkflowFromActivityRequest {
+  // instanceID is the orchestration instance ID to rerun.
+  string instanceID = 1;
+
+  // the event id to start the new workflow instance from.
+  int32 eventID = 2;
+
+  // newInstanceID is the optional new instance ID to use for the new workflow
+  // instance.
+  optional string newInstanceID = 3;
+
+  // input can optionally given to give the new instance a different input to
+  // the next Activity event.
+  google.protobuf.StringValue input = 4;
+}
+
+// RerunWorkflowFromActivityResponse is the response to executing
+// RerunWorkflowFromActivity.
+message RerunWorkflowFromActivityResponse {
+    string instanceId = 1;
+}
+```
+
+### Getting Instance History
+
+As a compliment to the `RerunWorkflowFromActivity` API, a new API is added to get the history of run activities for a workflow instance.
+Note that the API returns _all_ history events for the workflow instance, including control events which do _not_ contain an event ID.
+This API is intended to be used for discovering the event ID of the activity to rerun from.
+The actor backend will get the instance history from the state store and return it to the client, using a new workflow Actor invoke method.
+
+```proto
+service TaskHubSidecarService {
+    // GetInstanceHistory retrieves the history of a workflow instance.
+    rpc GetInstanceHistory(GetInstanceHistoryRequest) returns (GetInstanceHistoryResponse);
+}
+
+// RerunWorkflowFromActivityResponse is the response to executing
+// RerunWorkflowFromActivity.
+message RerunWorkflowFromActivityResponse {
+    string instanceId = 1;
+}
+
+// GetInstanceHistoryRequest is used to get the history of a workflow instance.
+message GetInstanceHistoryRequest {
+  // instanceID is the orchestration instance ID to get the history for.
+  string instanceID = 1;
+}
+
+// GetInstanceHistoryResponse is the response to executing
+// GetInstanceHistoryRequest.
+message GetInstanceHistoryResponse {
+  repeated HistoryEvent events = 1;
+}
+```
+
+### Concurrent Activities
+
+It is often the case that workflow activities are run concurrently, i.e. in fan-out patterns.
+This means the resulting workflow history order of execution can be non-deterministic.
+The durabletask history is currently a linear sequence of events.
+This then means that rerunning a workflow from a specific Activity which is a member of a fan-out pattern will result in possible rerunning of peer fan-out activities, depending on the order of termination of Activities in the fan-out group.
```

Comment:
I've been thinking if it's at all possible for the SDKs to actually force users to use an explicit Task.WhenAll etc. API instead of relying on the language primitives. That way the SDK can actually include metadata to tell the workflow engine that things are executed concurrently. This would be super useful for visualization. Or if there's some other magic way to do this then we could avoid this scenario?

Reply:
Actually we discussed looking at the history and if you have multiple TaskScheduled with Completion then you can infer they are concurrent?

Reply:
I'd be opposed to forcing users to use any particular library-specific logic as the shape, style and effect would differ by language. It's clear enough to say that the activity boundary is at the `await` in .NET whether it's a one-off activity or part of a `Task.WhenAll`.

Seems like you could fairly easily achieve your visualization idea using Cassie's event stream proposal - if you've got a bunch of activities that just kicked off and are running on a given workflow and you haven't yet received a completion for any, they were concurrently started.

Reply:
Not even theoretically - just because all the activities are started at the same time, there's absolutely no guarantee that they'd each run for the same length of time.